esx_service/utils/vmdk_utils.py

```diff
@@ -63,44 +63,57 @@
 VMDK_RETRY_COUNT = 5
 VMDK_RETRY_SLEEP = 1
 
-def init_datastoreCache():
+# For managing resource locks.
+lockManager = threadutils.LockManager()
+
+
+def init_datastoreCache(force=False):
     """
-    Initializes the datastore cache with the list of datastores accessible from local ESX host.
+    Initializes the datastore cache with the list of datastores accessible
+    from local ESX host. force=True will force it to ignore current cache
+    and force init
     """
     global datastores
     logging.debug("init_datastoreCache: %s", datastores)
 
-    si = vmdk_ops.get_si()
+    with lockManager.get_lock("init_datastoreCache"):
+        if datastores:
+            if force:
+                datastores = None
```

Comment:
this is what can cause issues, as @govint has mentioned ,on a non-locked consumption
it should simply say
```python
  if datastores and not force:
    return
```
and the code down the line should replaced datastores with tmp_datastores (local name bound to a new list)
And the last statement should simply be `datastores = tmp_datasores` which will re-bind `datastores` name to the new object. Note that the old object (list of datastores from the previous init()) may be in use by other threads at the same time and Python will simply garbage collects it when it's not used anymore